### PR TITLE
ENH: Compile Factorizer class for all numeric dtypes

### PR DESCRIPTION
**Pull request by phofl**
The current implementation has a couple of ugly side effects in merge:

- uint is cast to int, which causes overflow in some cases
- float is cast to object

This does not enable the classes yet. Will do as a follow up.

### CODE REVIEW

**Comment by phofl** [comment]
cc @jbrockmendel  would you mind having a look?

**Comment by phofl** [comment]
Adding factorize to the base class simplifies the follow up

File: pandas/_libs/hashtable.pyx [hunk 1]
```diff
@@ -69,17 +69,6 @@ else:
     raise ValueError(np.dtype(np.intp))
 
 
-cdef class Factorizer:
```
Comment (jbrockmendel) [inline pandas/_libs/hashtable.pyx:72]: this is removed here but still exists in the .pyi file?

Comment (jbrockmendel) [reply pandas/_libs/hashtable.pyx:72]: ... never mind, i see it is just moved below

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 2]
```diff
@@ -876,6 +890,44 @@ cdef class {{name}}HashTable(HashTable):
...
+cdef class {{name}}Factorizer(Factorizer):
+    cdef public:
+        {{name}}HashTable table
+        {{name}}Vector uniques
+
+    def __cinit__(self, size_hint: int):
+        self.table = {{name}}HashTable(size_hint)
+        self.uniques = {{name}}Vector()
+
+    def factorize(self, const {{c_type}}[:] values,
```
Comment (jbrockmendel) [inline pandas/_libs/hashtable_class_helper.pxi.in:903]: is the "const" here what keeps us from including ObjectFactorizer here?

**Comment by jbrockmendel** [comment]
> simplifies the follow up

i guess the follow-up will change how `klass` is found in merge.py?

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 3]
```diff
@@ -101,6 +101,20 @@ from pandas._libs.khash cimport (
 from pandas._libs.tslibs.util cimport get_c_string
 from pandas._libs.missing cimport C_NA
 
+
+cdef class Factorizer:
```
Comment (jbrockmendel) [inline pandas/_libs/hashtable_class_helper.pxi.in:105]: is there any particular reason for this to be in the pxi.in instead of the pyx?

**Comment by jbrockmendel** [comment]
a few questions, generally looks good

**Comment by phofl** [comment]
> > simplifies the follow up
> 
> i guess the follow-up will change how `klass` is found in merge.py?

Yeah and also avoids having a million typing errors

**Review by mroeschke — APPROVED** [review]
LGTM merge when ready @jbrockmendel

**Review by jbrockmendel — APPROVED** [review]

**Comment by jbrockmendel** [comment]
thanks @phofl

**Comment by rhshadrach** [comment]
This patch may have induced a potential regression. Please check the links below. If any ASVs are parameterized, the combinations of parameters that a regression has been detected appear as subbullets. This is a partially automated message.

 - https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsInForObjects.time_isin
   - series_type='long'; vals_type='nans'
 - https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsInLongSeriesLookUpDominates.time_isin
   - dtype='Float64'; MaxNumber=1000; series_type='monotone_hits'
   - dtype='float64'; MaxNumber=1000; series_type='monotone_hits'
   - dtype='object'; MaxNumber=1000; series_type='monotone_hits'
   - dtype='object'; MaxNumber=1000; series_type='random_misses'
   - dtype='object'; MaxNumber=5; series_type='monotone_hits'
 - https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsinWithArange.time_isin
   - dtype=<class 'numpy.object_'>; M=1000; offset_factor=-2
   - dtype=<class 'numpy.object_'>; M=2000; offset_factor=-2
   - dtype=<class 'numpy.object_'>; M=8000; offset_factor=-2
   - dtype=<class 'numpy.uint64'>; M=1000; offset_factor=-2
   - dtype=<class 'numpy.uint64'>; M=1000; offset_factor=0
   - dtype=<class 'numpy.uint64'>; M=2000; offset_factor=0
   - dtype=<class 'numpy.uint64'>; M=2000; offset_factor=2
   - dtype=<class 'numpy.uint64'>; M=8000; offset_factor=0

**Comment by phofl** [comment]
I think this is a false positive, looks to me as this is impacted by the klib quadratic probing and seems to go back to normal when this pr was reverted

**Comment by rhshadrach** [comment]
I'm seeing a few of the parametrizations showing regressions due to the revert of klib quadratic probing (as expected), but others point to this PR, e.g.

 - https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsInLongSeriesLookUpDominates.time_isin?p-MaxNumber=1000&p-series_type='monotone_hits'&p-dtype='object'
 - https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsInForObjects.time_isin?p-series_type='long'&p-vals_type='nans'

**Comment by phofl** [comment]
The links don't work. But this pr did not touch the code paths that the isin case takes

**Comment by rhshadrach** [comment]
Reworked my scripts and updated the links, the subbullets now contain links to single parameterization of the benchmark where my code is identifying a regression. But certainly could be a false positive. For example, the first subbullet of [algos.isin.IsInLongSeriesLookUpDominates.time_isin](https://asv-runner.github.io/asv-collection/pandas/#algos.isin.IsInLongSeriesLookUpDominates.time_isin) is an interesting case:

image

The first step up points to this PR whereas the second step up is the revert of quadratic programming. Both look like "regressions" but the revert of quadratic programming puts us back to where we started. I'm going to run asv's between certain commits; I'm just curious what's going on here.

**Pull request merged**